File: car_robot_ws/src/articubot/src/joy_serial.cpp
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/joy.hpp"
#include <serial/serial.h>
#include <string>
#include <sstream>

class JoyToSerial : public rclcpp::Node {
public:
    JoyToSerial() : Node("joy_to_serial"), current_speed(0) {
        // Initialize Serial Connection
        try {
            serial_port.setPort("/dev/ttyAMA0"); // Ensure this is the correct port
            serial_port.setBaudrate(115200);
            serial::Timeout timeout = serial::Timeout::simpleTimeout(1000);
            serial_port.setTimeout(timeout);
            serial_port.open();
        } catch (const std::exception &e) {
            RCLCPP_ERROR(this->get_logger(), "Failed to open serial port: %s", e.what());
        }

        // Subscribe to joystick topic
        subscription_ = this->create_subscription<sensor_msgs::msg::Joy>(
            "/joy", 10, std::bind(&JoyToSerial::joy_callback, this, std::placeholders::_1));

        // Steering and Speed parameters
        servo_min = 60;   // Left max
        servo_max = 120;  // Right max
        servo_center = 90; // Straight

        speed_min = 0;    // Min speed
        speed_max = 255;  // Max speed
    }

private:
    serial::Serial serial_port;
    rclcpp::Subscription<sensor_msgs::msg::Joy>::SharedPtr subscription_;

    int servo_min, servo_max, servo_center;
    int speed_min, speed_max, current_speed;

    void joy_callback(const sensor_msgs::msg::Joy::SharedPtr msg) {
        std::stringstream serial_command;

        // Get steering input from the Left Stick X-Axis
        float steering_input = msg->axes[0];  // Left (-1) to Right (+1)
        int servo_angle = static_cast<int>(servo_center + (steering_input * (servo_max - servo_center)));
        serial_command << "S" << servo_angle;

        // Get motor speed input from RT & LT buttons (Triggers)
        bool rt_pressed = msg->buttons[7];  // RT increases speed
        bool lt_pressed = msg->buttons[6];  // LT decreases speed

        if (rt_pressed) {
            current_speed = std::min(current_speed + 20, speed_max);  // Increase speed
        } else if (lt_pressed) {
            current_speed = std::max(current_speed - 20, speed_min);  // Decrease speed
        }

        serial_command << "M" << current_speed;

        // Button actions
        bool button_y = msg->buttons[4]; // Y for Forward
        bool button_a = msg->buttons[0]; // A for Backward
        bool button_x = msg->buttons[3]; // X for Halt
        // bool button_b = msg->buttons[2];

        if (button_y) {
            serial_command << "F";
        } else if (button_a) {
            serial_command << "B";
        } else if (button_x) {
            serial_command << "H"; // Halt
        }

        serial_command << "\n";

        // Send the command via Serial
        if (serial_port.isOpen()) {
            serial_port.write(serial_command.str());
        } else {
            RCLCPP_ERROR(this->get_logger(), "Serial port not open. Failed to send command.");
        }

        // Debugging logs
        RCLCPP_INFO(this->get_logger(), "Sent to Arduino: %s", serial_command.str().c_str());
    }
};

int main(int argc, char *argv[]) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<JoyToSerial>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: car_robot_ws/src/articubot/src/joy_serial.py
import rclpy
from rclpy.node import Node
import serial
from sensor_msgs.msg import Joy

class JoyToSerial(Node):
    def __init__(self):
        super().__init__('joy_to_serial')

        # Open serial port (Ensure correct port is used)
        self.serial_port = serial.Serial('/dev/ttyS0', baudrate=115200, timeout=1)

        # Subscribe to joystick topic
        self.subscription = self.create_subscription(
            Joy,
            '/joy',
            self.joy_callback,
            10
        )

        # Steering servo range (adjust if necessary)
        self.servo_min = 60   # Left max
        self.servo_max = 120  # Right max
        self.servo_center = 90  # Straight

        # Motor speed range
        self.speed_min = 0    # Min speed
        self.speed_max = 255  # Max speed
        self.current_speed = 0  # Initialize speed to 0

    def joy_callback(self, msg):
        serial_command = ""
        
        # Get steering input from the Left Stick X-Axis
        steering_input = msg.axes[0]  # Left (-1) to Right (+1)
        
        # Map joystick (-1 to 1) to servo angles (60° to 120°)
        servo_angle = int(self.servo_center + (steering_input * (self.servo_max - self.servo_center)))
        
        serial_command += f"S{servo_angle}"

        # Get motor speed input from RT & LT buttons (Triggers)
        rt_pressed = msg.buttons[5]  # RT increases speed
        lt_pressed = msg.buttons[4]  # LT decreases speed

        if rt_pressed:
            self.current_speed = min(self.current_speed + 20, self.speed_max)  # Increase speed
        elif lt_pressed:
            self.current_speed = max(self.current_speed - 20, self.speed_min)  # Decrease speed

        # Format the serial command as "SXXXMYYY\n" (Servo: 60-120, Motor: 0-255)
        serial_command += f"M{self.current_speed}"

        button_y = msg.buttons[3]
        button_a = msg.buttons[0]
        button_x = msg.buttons[2]

        if button_y:
            serial_command += "F"
        elif button_a:
            serial_command += "B"
        elif button_x
            serial_command += "H" #for halt

        serial_command += "\n"

        self.serial_port.write(serial_command.encode())

        # Debugging logs
        self.get_logger().info(f"Sent to Arduino: {serial_command}")

def main(args=None):
    rclpy.init(args=args)
    node = JoyToSerial()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
